File: Sources/Resources/registry.py
#!/usr/bin/env python3
# registry.py — app registry maintenance + brew-cask suggestion engine.
#
# Bridge extracted verbatim from update-all.sh's Phase-1 Python heredoc so the
# Swift RegistryTool can drive it while the logic is ported to Swift (Phase 4).
#
# Reads env: _STATE_DIR, _CACHE_DIR, _SCRIPT_DIR (optional), _BREW_SAFE_LIST
# (optional path; receives newline-separated "safe to install" cask tokens).
import json, os, subprocess
from pathlib import Path
from collections import Counter

SCRIPT_DIR = Path(os.environ.get('_SCRIPT_DIR', '/tmp'))
STATE_DIR  = Path(os.environ['_STATE_DIR'])
CACHE_DIR  = Path(os.environ['_CACHE_DIR'])
REGISTRY   = STATE_DIR / 'apps.json'
APPS_DIR   = Path('/Applications')

for _old in (SCRIPT_DIR / 'apps.json', SCRIPT_DIR.parent / 'apps.json'):
    if _old.exists() and not REGISTRY.exists():
        try: _old.rename(REGISTRY); print(f'  migrated apps.json → {REGISTRY}')
        except Exception: pass

DECLINED_FILE = STATE_DIR / 'declined-casks.json'
declined_casks = set()
if DECLINED_FILE.exists():
    try: declined_casks = set(json.loads(DECLINED_FILE.read_text()))
    except Exception: pass

def plist_get(plist, key):
    r = subprocess.run(['/usr/libexec/PlistBuddy', '-c', f'Print {key}', str(plist)],
                       capture_output=True, text=True)
    return r.stdout.strip() if r.returncode == 0 else None

def get_brew_apps():
    casks = subprocess.run(['brew', 'list', '--cask'], capture_output=True, text=True).stdout.split()
    if not casks: return {}
    r = subprocess.run(['brew', 'info', '--cask', '--json=v2'] + casks, capture_output=True, text=True)
    result = {}
    import os as _os
    try:
        for c in json.loads(r.stdout).get('casks', []):
            token = c['token']
            for art in c.get('artifacts', []):
                if not isinstance(art, dict): continue
                for name in art.get('app', []) or []:
                    if isinstance(name, str): result[name] = token
                for u in art.get('uninstall', []) or []:
                    if not isinstance(u, dict): continue
                    for path in u.get('delete', []) or []:
                        if not isinstance(path, str): continue
                        if not (path.startswith('/Applications/') and path.endswith('.app')): continue
                        if '*' in path: continue
                        result[_os.path.basename(path)] = token
    except Exception: pass
    return result

def get_mas_apps():
    r = subprocess.run(['mas', 'list'], capture_output=True, text=True)
    result = {}
    for line in r.stdout.strip().splitlines():
        parts = line.strip().split(None, 1)
        if len(parts) == 2:
            mas_id, rest = parts
            result[f'{rest.rsplit("(", 1)[0].strip()}.app'] = mas_id
    return result

def load_cask_catalog():
    catalog_path = CACHE_DIR / 'brew_cask_catalog_v4.json'
    import time, os, sys
    if catalog_path.exists() and time.time() - catalog_path.stat().st_mtime < 86400:
        return json.loads(catalog_path.read_text())
    print('  ⤓ Fetching Homebrew cask catalog (~15MB, cached for 24h)...', flush=True)
    r = subprocess.run(['curl', '-sf', 'https://formulae.brew.sh/api/cask.json'],
                       capture_output=True, text=True)
    if r.returncode != 0:
        print('    ✗ catalog fetch failed — continuing without cask suggestions')
        return {}
    casks = json.loads(r.stdout)
    catalog = {}
    for c in casks:
        token, ver = c['token'], c.get('version', '')
        apps_for_this = set()
        for art in c.get('artifacts', []):
            if not isinstance(art, dict): continue
            for a in art.get('app', []) or []:
                if isinstance(a, str): apps_for_this.add(a)
            for u in art.get('uninstall', []) or []:
                if not isinstance(u, dict): continue
                for path in u.get('delete', []) or []:
                    if not isinstance(path, str): continue
                    if not (path.startswith('/Applications/') and path.endswith('.app')): continue
                    if '*' in path: continue
                    name = os.path.basename(path)
                    if name: apps_for_this.add(name)
        for a in apps_for_this:
            catalog.setdefault(a.lower().removesuffix('.app'), []).append({
                'token': token, 'version': ver,
            })
    catalog_path.write_text(json.dumps(catalog))
    print(f'    ✓ catalog indexed: {len(catalog)} apps from {len(casks)} casks', flush=True)
    for old in (CACHE_DIR / 'brew_cask_catalog.json',
                CACHE_DIR / 'brew_cask_catalog_v2.json',
                CACHE_DIR / 'brew_cask_catalog_v3.json',
                SCRIPT_DIR / 'brew_cask_catalog.json',
                SCRIPT_DIR / 'brew_cask_catalog_v2.json'):
        if old.exists():
            try: old.unlink()
            except Exception: pass
    return catalog

_cask_catalog = None
def find_brew_cask(app_name, installed_version=None):
    global _cask_catalog
    if _cask_catalog is None:
        _cask_catalog = load_cask_catalog()
    candidates = _cask_catalog.get(app_name.lower().removesuffix('.app'))
    if not candidates: return None
    if len(candidates) == 1: return candidates[0]
    def major(s):
        import re as _re
        m = _re.match(r'(\d+)', s or '')
        return int(m.group(1)) if m else -1
    if installed_version:
        iv_maj = major(installed_version)
        same = [c for c in candidates if major(c.get('version', '')) == iv_maj]
        if same: return same[0]
    return max(candidates, key=lambda c: major(c.get('version', '')))

import re as _re
def _first_int(s):
    m = _re.match(r'(\d+)', s or '')
    return int(m.group(1)) if m else None

def _sortv(a, b):
    if a == b: return 0
    r = subprocess.run(['bash', '-c', f'printf "%s\\n%s\\n" "{a}" "{b}" | sort -V | tail -1'],
                       capture_output=True, text=True)
    top = r.stdout.strip()
    return 1 if top == a else (-1 if top == b else 0)

def cmp_version(a, b):
    if not a or not b: return None
    a_clean = a.split(',')[0].split()[0]
    b_clean = b.split(',')[0].split()[0]
    if a_clean == b_clean: return 0
    fa, fb = _first_int(a_clean), _first_int(b_clean)
    pairs = [(a_clean, b_clean)]
    if fa is not None and fb is not None:
        if fa >= 10 * max(fb, 1) and '.' in a_clean:
            pairs.append((a_clean.split('.', 1)[1], b_clean))
        if fb >= 10 * max(fa, 1) and '.' in b_clean:
            pairs.append((a_clean, b_clean.split('.', 1)[1]))
    for ax, bx in pairs:
        ma, mb = _first_int(ax), _first_int(bx)
        if ma is None or mb is None: continue
        hi, lo = max(ma, mb), max(min(ma, mb), 1)
        if hi < 10 * lo:
            return _sortv(ax, bx)
    return None

brew_managed = get_brew_apps()
mas_managed  = get_mas_apps()
managed      = set(brew_managed) | set(mas_managed)
installed    = {p.name for p in APPS_DIR.glob('*.app')}
unmanaged    = installed - managed

print(f'  {len(installed)} apps in /Applications')
print(f'  brew: {len(brew_managed)}  mas: {len(mas_managed)}  unmanaged: {len(unmanaged)}')

registry = json.loads(REGISTRY.read_text()) if REGISTRY.exists() else {}
changed  = False

for name in [n for n in list(registry) if n in managed or n not in installed]:
    print(f'  - {name} (now managed or removed)')
    del registry[name]; changed = True

brew_suggestions = []
for name in sorted(unmanaged - set(registry)):
    plist = APPS_DIR / name / 'Contents' / 'Info.plist'
    if plist.exists():
        feed = plist_get(plist, 'SUFeedURL')
        if feed:
            entry = {'manager': 'sparkle', 'feed_url': feed}
        elif (APPS_DIR / name / 'Contents/Frameworks/Electron Framework.framework').exists():
            entry = {'manager': 'electron'}
        else:
            entry = {'manager': 'unmanaged'}
    else:
        entry = {'manager': 'unmanaged'}
    installed_ver = plist_get(plist, 'CFBundleShortVersionString') if plist.exists() else ''
    cask = find_brew_cask(name, installed_ver)
    if cask:
        entry['brew_cask_available'] = cask['token']
        brew_suggestions.append((name, cask['token']))
    print(f'  + {name}  [{entry["manager"]}]{f"  → brew cask: {cask["token"]}" if cask else ""}')
    registry[name] = entry; changed = True

if changed:
    REGISTRY.write_text(json.dumps(registry, indent=2, sort_keys=True))
    print(f'  Saved → {REGISTRY}')
else:
    print(f'  Up to date')

counts = Counter(e['manager'] for e in registry.values())
print(f'  sparkle: {counts.get("sparkle",0)}  electron: {counts.get("electron",0)}  unmanaged: {counts.get("unmanaged",0)}')

def _grouped(predicate):
    return sorted(n[:-4] for n, e in registry.items() if predicate(e))
_sparkle  = _grouped(lambda e: e['manager'] == 'sparkle')
_electron = _grouped(lambda e: e['manager'] == 'electron')
_brewable = _grouped(lambda e: e.get('brew_cask_available'))
_orphan   = _grouped(lambda e: e['manager'] == 'unmanaged' and not e.get('brew_cask_available'))
if _sparkle:
    print(f'\n  Sparkle (self-update on launch):')
    for n in _sparkle: print(f'    • {n}')
if _electron:
    print(f'\n  Electron (self-update on launch):')
    for n in _electron: print(f'    • {n}')
if _brewable:
    print(f'\n  Brew cask available (offered below):')
    for n in _brewable: print(f'    • {n}')
if _orphan:
    print(f'\n  Unmanaged — no auto-update detected (check manually):')
    for n in _orphan: print(f'    • {n}')

safe = []; downgrade = []; mismatch = []; unknown = []
for n, e in registry.items():
    plist = APPS_DIR / n / 'Contents' / 'Info.plist'
    installed_ver = plist_get(plist, 'CFBundleShortVersionString') if plist.exists() else ''
    cask_info = find_brew_cask(n, installed_ver) or {}
    new_token = cask_info.get('token')
    cur_token = e.get('brew_cask_available')
    if new_token != cur_token:
        if new_token:
            e['brew_cask_available'] = new_token; changed = True
            verb = 'added' if not cur_token else 'updated'
            print(f'  ⟳ cask mapping {verb}: {n[:-4]} → {new_token}')
        elif cur_token:
            del e['brew_cask_available']; changed = True
            print(f'  ⟳ cask mapping removed: {n[:-4]} (was {cur_token})')
    if not new_token: continue
    token = new_token
    if token in declined_casks: continue
    brew_ver = cask_info.get('version', '')
    if not brew_ver or not installed_ver:
        unknown.append((n, token, brew_ver, installed_ver))
    else:
        c = cmp_version(brew_ver, installed_ver)
        if c is None:
            mismatch.append((n, token, brew_ver, installed_ver))
        elif c >= 0:
            safe.append((n, token, brew_ver, installed_ver))
        else:
            downgrade.append((n, token, brew_ver, installed_ver))

if safe:
    print(f'\n  💡 Brew cask available (safe — same or newer):')
    by_token = {}
    for n, t, bv, iv in sorted(safe):
        by_token.setdefault(t, []).append((n, bv, iv))
    for t in sorted(by_token):
        items = by_token[t]
        if len(items) == 1:
            n, bv, iv = items[0]
            tag = f'{iv}' if bv == iv else f'{iv} → {bv}'
            print(f'     {n[:-4]}  ({tag})  →  brew install --cask --force {t}')
        else:
            print(f'     {t}  (bundle — covers {len(items)} apps)  →  brew install --cask --force {t}')
            for n, bv, iv in items:
                tag = f'{iv}' if bv == iv else f'{iv} → {bv}'
                print(f'        • {n[:-4]}  ({tag})')

if downgrade:
    print(f'\n  ⚠  Brew has older version (would downgrade — skipped):')
    for n, t, bv, iv in sorted(downgrade):
        print(f'     {n[:-4]}: installed {iv}, brew {bv}')

if mismatch:
    print(f'\n  ≠  Version schemes differ — different cask variant or product? (skipped):')
    for n, t, bv, iv in sorted(mismatch):
        print(f'     {n[:-4]}: installed {iv}, brew "{t}" reports {bv}')

if unknown:
    print(f'\n  ?  Brew cask available (version unknown):')
    for n, t, bv, iv in sorted(unknown):
        print(f'     {n[:-4]}  →  brew install --cask --force {t}')

if not (safe or downgrade or mismatch or unknown):
    print(f'\n  ✓ No unmanaged apps have a matching brew cask — nothing to suggest')

if changed:
    REGISTRY.write_text(json.dumps(registry, indent=2, sort_keys=True))

safe_file = os.environ.get('_BREW_SAFE_LIST', '')
if safe_file and safe:
    seen = set()
    with open(safe_file, 'w') as f:
        for n, t, bv, iv in safe:
            if t in seen: continue
            seen.add(t)
            f.write(t + '\n')

File: Sources/Core/Registry.swift
import Foundation

/// One tracked unmanaged app in apps.json. Property names match the JSON keys
/// (snake_case) so Codable round-trips without custom keys; nil optionals are
/// omitted on encode (synthesized encodeIfPresent).
struct RegistryEntry: Codable {
    var manager: String                 // "sparkle" | "electron" | "unmanaged"
    var feed_url: String?               // sparkle appcast URL
    var brew_cask_available: String?    // matching cask token, if any
    var last_version: String?           // last version we installed via Sparkle
}

/// apps.json maintenance + brew-cask suggestion engine. Native port of
/// registry.py. All methods are synchronous (Process-blocking) and meant to be
/// called from a tool's off-main async context.
enum Registry {
    static func load() -> [String: RegistryEntry] {
        guard let data = FileManager.default.contents(atPath: AppPaths.registry) else { return [:] }
        return (try? JSONDecoder().decode([String: RegistryEntry].self, from: data)) ?? [:]
    }

    static func save(_ r: [String: RegistryEntry]) {
        let enc = JSONEncoder()
        enc.outputFormatting = [.prettyPrinted, .sortedKeys]
        if let data = try? enc.encode(r) {
            try? data.write(to: URL(fileURLWithPath: AppPaths.registry))
        }
    }

    static func declinedCasks() -> Set<String> {
        guard let data = FileManager.default.contents(atPath: AppPaths.state + "/declined-casks.json"),
              let arr = try? JSONSerialization.jsonObject(with: data) as? [String] else { return [] }
        return Set(arr)
    }

    /// /Applications/*.app names → cask token, for every installed cask
    /// (reading both `app:` artifacts and uninstall.delete paths).
    static func brewManagedApps() -> [String: String] {
        let casks = Shell.capture(["brew", "list", "--cask"]).out
            .split(whereSeparator: { $0 == " " || $0 == "\n" }).map(String.init)
        guard !casks.isEmpty else { return [:] }
        let r = Shell.capture(["brew", "info", "--cask", "--json=v2"] + casks)
        guard let data = r.out.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let arr = obj["casks"] as? [[String: Any]] else { return [:] }
        var result: [String: String] = [:]
        for c in arr {
            guard let token = c["token"] as? String else { continue }
            for art in (c["artifacts"] as? [[String: Any]]) ?? [] {
                for n in (art["app"] as? [Any]) ?? [] { if let s = n as? String { result[s] = token } }
                for u in (art["uninstall"] as? [[String: Any]]) ?? [] {
                    for p in (u["delete"] as? [Any]) ?? [] {
                        if let path = p as? String, path.hasPrefix("/Applications/"),
                           path.hasSuffix(".app"), !path.contains("*") {
                            result[(path as NSString).lastPathComponent] = token
                        }
                    }
                }
            }
        }
        return result
    }

    /// "App Name.app" → MAS id, for every Mac App Store app.
    static func masManagedApps() -> [String: String] {
        let out = Shell.capture(["mas", "list"]).out
        var result: [String: String] = [:]
        for line in out.split(separator: "\n") {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            guard let firstSpace = trimmed.firstIndex(of: " ") else { continue }
            let rest = trimmed[trimmed.index(after: firstSpace)...].trimmingCharacters(in: .whitespaces)
            // strip trailing " (version)"
            let name: String
            if let paren = rest.range(of: "(", options: .backwards) {
                name = String(rest[..<paren.lowerBound]).trimmingCharacters(in: .whitespaces)
            } else {
                name = rest
            }
            if !name.isEmpty { result["\(name).app"] = String(trimmed[..<firstSpace]) }
        }
        return result
    }

    /// Full registry sync + categorization. Emits the same human output the old
    /// registry.py printed, updates apps.json, and returns the deduped list of
    /// "safe" (same-or-newer) cask tokens to offer for install.
    @discardableResult
    static func sync(catalog: CaskCatalog, emit: (String) -> Void) -> [String] {
        let appsDir = "/Applications"
        let fm = FileManager.default
        let declined = declinedCasks()

        func plist(_ app: String) -> String { "\(appsDir)/\(app)/Contents/Info.plist" }
        func installedVer(_ app: String) -> String { Plist.value(plist(app), "CFBundleShortVersionString") ?? "" }

        let brewManaged = brewManagedApps()
        let masManaged  = masManagedApps()
        let managed = Set(brewManaged.keys).union(masManaged.keys)
        let installed = Set((try? fm.contentsOfDirectory(atPath: appsDir))?.filter { $0.hasSuffix(".app") } ?? [])
        let unmanaged = installed.subtracting(managed)

        emit("  \(installed.count) apps in /Applications\n")
        emit("  brew: \(brewManaged.count)  mas: \(masManaged.count)  unmanaged: \(unmanaged.count)\n")

        var registry = load()
        var changed = false

        // prune entries now managed or removed
        for name in registry.keys where managed.contains(name) || !installed.contains(name) {
            emit("  - \(name) (now managed or removed)\n")
            registry[name] = nil; changed = true
        }

        // classify newly-seen unmanaged apps
        for name in unmanaged.subtracting(registry.keys).sorted() {
            let p = plist(name)
            let hasPlist = fm.fileExists(atPath: p)
            var entry: RegistryEntry
            if hasPlist, let feed = Plist.value(p, "SUFeedURL") {
                entry = RegistryEntry(manager: "sparkle", feed_url: feed)
            } else if fm.fileExists(atPath: "\(appsDir)/\(name)/Contents/Frameworks/Electron Framework.framework") {
                entry = RegistryEntry(manager: "electron")
            } else {
                entry = RegistryEntry(manager: "unmanaged")
            }
            let cask = catalog.find(name, installedVersion: hasPlist ? installedVer(name) : nil)
            if let cask = cask { entry.brew_cask_available = cask.token }
            emit("  + \(name)  [\(entry.manager)]\(cask.map { "  → brew cask: \($0.token)" } ?? "")\n")
            registry[name] = entry; changed = true
        }

        if changed { save(registry); emit("  Saved → \(AppPaths.registry)\n") }
        else { emit("  Up to date\n") }

        let mgrCount = { (m: String) in registry.values.filter { $0.manager == m }.count }
        emit("  sparkle: \(mgrCount("sparkle"))  electron: \(mgrCount("electron"))  unmanaged: \(mgrCount("unmanaged"))\n")

        func grouped(_ pred: (RegistryEntry) -> Bool) -> [String] {
            registry.filter { pred($0.value) }.keys.map { $0.removingSuffix(".app") }.sorted()
        }
        let sparkle  = grouped { $0.manager == "sparkle" }
        let electron = grouped { $0.manager == "electron" }
        let brewable = grouped { $0.brew_cask_available != nil }
        let orphan   = grouped { $0.manager == "unmanaged" && $0.brew_cask_available == nil }
        func list(_ title: String, _ names: [String]) {
            guard !names.isEmpty else { return }
            emit("\n  \(title)\n"); for n in names { emit("    • \(n)\n") }
        }
        list("Sparkle (self-update on launch):", sparkle)
        list("Electron (self-update on launch):", electron)
        list("Brew cask available (offered below):", brewable)
        list("Unmanaged — no auto-update detected (check manually):", orphan)

        // categorize cask matches by version comparison
        typealias Row = (name: String, token: String, brewVer: String, instVer: String)
        var safe: [Row] = [], downgrade: [Row] = [], mismatch: [Row] = [], unknown: [Row] = []
        for (name, var entry) in registry {
            let iv = installedVer(name)
            let cask = catalog.find(name, installedVersion: iv)
            let newToken = cask?.token
            if newToken != entry.brew_cask_available {
                if let t = newToken {
                    let verb = entry.brew_cask_available == nil ? "added" : "updated"
                    entry.brew_cask_available = t; registry[name] = entry; changed = true
                    emit("  ⟳ cask mapping \(verb): \(name.removingSuffix(".app")) → \(t)\n")
                } else if let old = entry.brew_cask_available {
                    entry.brew_cask_available = nil; registry[name] = entry; changed = true
                    emit("  ⟳ cask mapping removed: \(name.removingSuffix(".app")) (was \(old))\n")
                }
            }
            guard let token = newToken else { continue }
            if declined.contains(token) { continue }
            let bv = cask?.version ?? ""
            if bv.isEmpty || iv.isEmpty { unknown.append((name, token, bv, iv)); continue }
            switch Version.compare(bv, iv) {
            case .none:           mismatch.append((name, token, bv, iv))
            case .some(let c) where c >= 0: safe.append((name, token, bv, iv))
            default:              downgrade.append((name, token, bv, iv))
            }
        }

        if !safe.isEmpty {
            emit("\n  💡 Brew cask available (safe — same or newer):\n")
            var byToken: [String: [Row]] = [:]
            for r in safe.sorted(by: { $0.name < $1.name }) { byToken[r.token, default: []].append(r) }
            for t in byToken.keys.sorted() {
                let items = byToken[t]!
                if items.count == 1 {
                    let r = items[0]
                    let tag = r.brewVer == r.instVer ? r.instVer : "\(r.instVer) → \(r.brewVer)"
                    emit("     \(r.name.removingSuffix(".app"))  (\(tag))  →  brew install --cask --force \(t)\n")
                } else {
                    emit("     \(t)  (bundle — covers \(items.count) apps)  →  brew install --cask --force \(t)\n")
                    for r in items {
                        let tag = r.brewVer == r.instVer ? r.instVer : "\(r.instVer) → \(r.brewVer)"
                        emit("        • \(r.name.removingSuffix(".app"))  (\(tag))\n")
                    }
                }
            }
        }
        if !downgrade.isEmpty {
            emit("\n  ⚠  Brew has older version (would downgrade — skipped):\n")
            for r in downgrade.sorted(by: { $0.name < $1.name }) {
                emit("     \(r.name.removingSuffix(".app")): installed \(r.instVer), brew \(r.brewVer)\n")
            }
        }
        if !mismatch.isEmpty {
            emit("\n  ≠  Version schemes differ — different cask variant or product? (skipped):\n")
            for r in mismatch.sorted(by: { $0.name < $1.name }) {
                emit("     \(r.name.removingSuffix(".app")): installed \(r.instVer), brew \"\(r.token)\" reports \(r.brewVer)\n")
            }
        }
        if !unknown.isEmpty {
            emit("\n  ?  Brew cask available (version unknown):\n")
            for r in unknown.sorted(by: { $0.name < $1.name }) {
                emit("     \(r.name.removingSuffix(".app"))  →  brew install --cask --force \(r.token)\n")
            }
        }
        if safe.isEmpty && downgrade.isEmpty && mismatch.isEmpty && unknown.isEmpty {
            emit("\n  ✓ No unmanaged apps have a matching brew cask — nothing to suggest\n")
        }

        if changed { save(registry) }

        // deduped safe tokens, order preserved
        var seen = Set<String>(), tokens: [String] = []
        for r in safe where !seen.contains(r.token) { seen.insert(r.token); tokens.append(r.token) }
        return tokens
    }

    /// Append declined cask tokens so they aren't re-offered.
    static func addDeclines(_ tokens: [String]) {
        var set = declinedCasks()
        set.formUnion(tokens)
        if let data = try? JSONSerialization.data(withJSONObject: set.sorted(), options: [.prettyPrinted]) {
            try? data.write(to: URL(fileURLWithPath: AppPaths.state + "/declined-casks.json"))
        }
    }
}
